jax/_src/scipy/stats/_core.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2022 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+from functools import partial
+from typing import Optional, Tuple
+
+import jax.numpy as jnp
+import scipy
+from jax import jit
+from jax._src.api import vmap
+from jax._src.numpy.lax_numpy import _check_arraylike
+from jax._src.numpy.util import _wraps
+from jax._src.typing import ArrayLike
+from jax._src.util import canonicalize_axis
+
+ModeResult = namedtuple('ModeResult', ('mode', 'count'))
+
+@_wraps(scipy.stats.mode, lax_description="""\
+Currently the only supported nan_policy is 'propagate'
+""")
+@partial(jit, static_argnames=['axis', 'nan_policy', 'keepdims'])
+def mode(a: ArrayLike, axis: Optional[int] = 0, nan_policy: str = "propagate", keepdims: bool = False) -> ModeResult:
+  _check_arraylike("mode", a)
+  x = jnp.atleast_1d(a)
+
+  if nan_policy not in ["propagate", "omit", "raise"]:
+    raise ValueError(
+      f"Illegal nan_policy value {nan_policy!r}; expected one of "
+      "{'propoagate', 'omit', 'raise'}"
+    )
+  if nan_policy == "omit":
+    # TODO: return answer without nans included.
+    raise NotImplementedError(
+      f"Logic for `nan_policy` of {nan_policy} is not implemented"
+    )
+  if nan_policy == "raise":
+    raise NotImplementedError(
+      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
+      "Please check if nans exist in `x` outside of the `mode` function."
+    )
+
+  if x.size == 0:
+    if axis is None:
+      if keepdims:
+        output_shape = [1 for _ in range(x.ndim)]
+        return ModeResult(jnp.full(output_shape, jnp.nan), jnp.full(output_shape, jnp.nan))
+      else:
+        return ModeResult(jnp.nan, jnp.nan)
```

Comment:
`jnp.nan` is a python scalar `NaN`... for consistency we should return a jax array in all cases. We should use `jnp.full` in both cases... maybe compute the output_shape within the `if` statement and only have a single return statement? Maybe even one `return` statement for the entire `if x.size == 0` block? The only thing that's changing is the output shape, yes?

Reply:
Also this is a bit in the weeds, but `jnp.full(shape, jnp.nan)` returns a weakly-typed array because `jnp.nan` is a Python scalar. For consistency with the non-empty return value we should use
```python
jnp.full(shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))
```

Reply:
Another thought... this may all be cleaner if you handle the empty case in `_mode_helper` instead.

Reply:
```scipy.stats.mode(np.zeros((0,)), axis=None, keepdims=False)``` and ```scipy.stats.mode(np.zeros((0,)), axis=0, keepdims=False)``` both return ```ModeResult(mode=nan, count=nan)```

Reply:
But they don't return a Python scalar NaN, they return a numpy.float64 NaN (i.e. a numpy scalar). JAX's equivalent of a numpy scalar is a zero-dimensional array.

Reply:
This is important because, for example, someone might write code using something like `result.mode.shape`, and if we return a Python scalar NaN in some cases, that would result in an AttributeError. For that reason we generally try to keep return types consistent for different shape inputs. Does that make sense?

Reply:
Yes it makes sense.
So instead of returning ```ModeResult(jnp.nan, jnp.nan)``` in this case, I can return  
```ModeResult(jnp.full((), jnp.nan), jnp.full((), jnp.nan))``` 
?

Reply:
Yes, basically, except we should specify the dtype: `jnp.full(shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))`

Reply:
To make the code a bit cleaner, I can refactor it like this:
```python
  def _mode_result_nan(output_shape):
    return ModeResult(
      jnp.full(output_shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)),
      jnp.full(output_shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)))

  def _get_output_shape(shape):
    output_shape = list(shape)
    if keepdims:
      output_shape[axis] = 1
    else:
      del output_shape[axis]
    return output_shape

  if x.size == 0:
    if axis is None:
      output_shape = [1 for _ in range(x.ndim)] if keepdims else ()
      return _mode_result_nan(output_shape)
    else:
      return _mode_result_nan(_get_output_shape(x.shape))
```

Reply:
I'd avoid that kind of extra indirection in the form of one-line helper functions. I was just saying that you might define `output_shape` within the `if`/`else` blocks and then have a single return statement at the end. But if you prefer multiple return statements, that's fine too.

Reply:
I think the best approach would be to handle the empty case within the vmapped helper, but we've already been going a long time here and I don't want to make you refactor what you have too much!

Reply:
So i added a helper function and pushed the changes.   
I'm not sure that's exactly what you wanted for the helper function.

Reply:
Sorry, I think I was unclear. I don't think we need all these extra helper functions for this algorithm: it's too much indirection and makes the code much harder to follow.

When I said "handle the empty case within the vmapped helper", I meant something like this:
```python
  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
    """Helper function to return mode and count of a given array."""
    if x.size == 0:
      return jnp.float32(jnp.nan), jnp.float32(jnp.nan)
    else:
      vals, counts = jnp.unique(x, return_counts=True, size=x.size)
      return vals[jnp.argmax(counts)], counts.max()
```
Then I think the zero-size case would be handled naturally within the `vmap` where this function is applied.

Reply:
Yes, but then I have error in the following cases:
```python
shape=(0,), axis=None, keepdims=False
shape=(0,), axis=None, keepdims=True
shape=(0,), axis=0, keepdims=False
shape=(0,), axis=0, keepdims=True
shape=(0, 2, 3), axis=None, keepdims=False
shape=(0, 2, 3), axis=None, keepdims=True
shape=(0, 2, 3), axis=0, keepdims=False
shape=(0, 2, 3), axis=0, keepdims=True
```

Reply:
Without seeing the error, I suspect it's coming from `x = x.reshape(x.shape[0], -1)` which is incompatible with zero-sized arrays. you can do `x = x.reshape(x.shape[0], prod(x.shape[1:]))` instead and that will work for arrays of any size

Reply:
Yes, thanks. But I also have errors like:
```python
ValueError: vmap was requested to map its argument along axis 1, which implies that its rank should be at least 2, but is only 1 (its shape is (0,))
```

Reply:
Hard to guess without seeing the code. It sound like you removed the `reshape` line if your array has only one dimension when you pass it to the vmapped function

Reply:
Sorry, my mistake. There is another error:
```
jax._src.traceback_util.UnfilteredStackTrace: UnboundLocalError: local variable 'vals' referenced before assignment
```

Reply:
Something like this should work:
```python
def mode(a: ArrayLike, axis: Optional[int] = 0, nan_policy: str = "propagate", keepdims: bool = False) -> ModeResult:
  _check_arraylike("mode", a)
  x = jnp.atleast_1d(a)

  if nan_policy not in ["propagate", "omit", "raise"]:
    raise ValueError(
      f"Illegal nan_policy value {nan_policy!r}; expected one of "
      "{'propoagate', 'omit', 'raise'}"
    )
  if nan_policy == "omit":
    # TODO: return answer without nans included.
    raise NotImplementedError(
      f"Logic for `nan_policy` of {nan_policy} is not implemented"
    )
  if nan_policy == "raise":
    raise NotImplementedError(
      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
      "Please check if nans exist in `x` outside of the `mode` function."
    )

  input_shape = x.shape
  if keepdims:
    if axis is None:
      output_shape = tuple(1 for i in input_shape)
    else:
      output_shape = tuple(1 if i == axis else s for i, s in enumerate(input_shape))
  else:
    if axis is None:
      output_shape = ()
    else:
      output_shape = tuple(s for i, s in enumerate(input_shape) if i != axis)

  if axis is None:
    axis = 0
    x = x.ravel()

  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
    """Helper function to return mode and count of a given array."""
    if x.size == 0:
      return jnp.array(jnp.nan, dtype=x.dtype), jnp.float32(jnp.nan)
    else:
      vals, counts = jnp.unique(x, return_counts=True, size=x.size)
      return vals[jnp.argmax(counts)], counts.max()

  axis = canonicalize_axis(axis, x.ndim)
  x = jnp.moveaxis(x, axis, 0)
  x = x.reshape(x.shape[0], prod(x.shape[1:]))
  vals, counts = vmap(_mode_helper, in_axes=1)(x)
  return ModeResult(vals.reshape(output_shape), counts.reshape(output_shape))
```
It won't match scipy for `keepdims=True` and `axis=None`, but scipy is wrong in that case.

Reply:
I just realized we can't return `jnp.array(jnp.nan, dtype=x.dtype)` because `x` may be integer typed. I'm not sure what's best in that case.

If you want to go back to what you had yesterday (before the helper function refactor) I'd be fine with that too. Your call

Reply:
But we should make sure we *don't* match scipy for `keepdims=True` and `axis=None`. The return shape in that case should be `(1,) * len(input_shape)`

Reply:
Just to clarify:
```python
scipy.stats.mode(np.ones((0, 3, 2)), axis=None, keepdims=True).mode.shape
```
returns ```(1, 1, 1)```.
And
```python
scipy.stats.mode(np.ones((4, 3, 2)), axis=None, keepdims=True).mode.shape
```
returns ```(1,)```.

So, ```jax.scipy.stats.mode``` should replicate ```scipy.stats.mode``` in both cases ?

Reply:
Sorry, I mixed that up. I think it should return `(1, 1, 1)` in both cases. From https://github.com/scipy/scipy/issues/17214#issuecomment-1276464112 it's my understanding that this is the intended API.

Reply:
Ok. 
But then the tests where x is not empty, ```ndim>=2 , axis=None, keepdims=True``` will fail because the shape returned by ```jax.scipy.stats.mode``` is not the same as the shaped returned by ```scipy.stats.mode```.
Is there a way to deactivate the test on the shape in these cases ?

Reply:
Another thing. Shouldn't ```_mode_helper``` return 
```python
return jnp.array(jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)), jnp.array(jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))
```
when ```x.size== 0```

Reply:
> Is there a way to deactivate the test on the shape in these cases ?

I'd probably approach this by writing a wrapper around the scipy function that detects this case and returns the expected result.

> Shouldn't _mode_helper return...

Yes, you're right

Reply:
Ok. I made the changes and pushed the result